build/get_concurrent_jobs.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+# Copyright 2019 The Fuchsia Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# This script computes the number of concurrent jobs that can run in the
+# build as a function of the machine.
+
+import argparse
+import ctypes
+import json
+import multiprocessing
+import os
+import re
+import subprocess
+import sys
+
+UNITS = {'B': 1, 'KB': 2**10, 'MB': 2**20, 'GB': 2**30, 'TB': 2**40}
+
+
+# See https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex
+# and https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-memorystatusex
+class MEMORYSTATUSEX(ctypes.Structure):
+  _fields_ = [
+      ("dwLength", ctypes.c_ulong),
+      ("dwMemoryLoad", ctypes.c_ulong),
+      ("ullTotalPhys", ctypes.c_ulonglong),
+      ("ullAvailPhys", ctypes.c_ulonglong),
+      ("ullTotalPageFile", ctypes.c_ulonglong),
+      ("ullAvailPageFile", ctypes.c_ulonglong),
+      ("ullTotalVirtual", ctypes.c_ulonglong),
+      ("ullAvailVirtual", ctypes.c_ulonglong),
+      ("sullAvailExtendedVirtual", ctypes.c_ulonglong),
+  ]
+
+
+def GetTotalMemory():
+  if sys.platform in ('win32', 'cygwin'):
+    stat = MEMORYSTATUSEX(dwLength=ctypes.sizeof(MEMORYSTATUSEX))
+    ctypes.windll.kernel32.GlobalMemoryStatusEx(ctypes.byref(stat))
+    return stat.ullTotalPhys
+  elif sys.platform.startswith('linux'):
+    if os.path.exists("/proc/meminfo"):
+      with open("/proc/meminfo") as meminfo:
+        memtotal_re = re.compile(r'^MemTotal:\s*(\d*)\s*kB')
+        for line in meminfo:
+          match = memtotal_re.match(line)
+          if match:
+            return float(match.group(1)) * 2**10
+  elif sys.platform == 'darwin':
+    try:
+      return int(subprocess.check_output(['sysctl', '-n', 'hw.memsize']))
+    except Exception:
+      return 0
+  else:
+    return 0
+
+
+def ParseSize(string):
+  i = next(i for (i, c) in enumerate(string) if not c.isdigit())
+  number = string[:i].strip()
+  unit = string[i:].strip()
+  return int(float(number) * UNITS[unit])
+
+
+class ParseSizeAction(argparse.Action):
+
+  def __call__(self, parser, args, values, option_string=None):
+    sizes = getattr(args, self.dest, [])
+    for value in values:
+      (k, v) = value.split('=', 1)
+      sizes.append((k, ParseSize(v)))
+    setattr(args, self.dest, sizes)
+
+
+def Main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+      "--memory-per-job", action=ParseSizeAction, default=[], nargs='*')
+    parser.add_argument("--reserve-memory", type=ParseSize, default=0)
+    args = parser.parse_args()
+
+    total_memory = GetTotalMemory()
```

Comment:
If total_memory is zero (as it could be in the error cases), perhaps just quit and return a default json with job counts of 1? I am not sure how robust GetTotalMemory is to error, or (for that matter) GN is to zero job counts.

Reply:
Added comments.